Issue #6882: REGEXP_EXTRACT Capture Groups

hawkfish opened the pull request:
Add a new variant that takes a list of field names and produces a STRUCT of all the capture groups.

Commit:
Issue #6882: REGEXP_EXTRACT Capture Groups

Add a new variant that takes a list of field names and produces a STRUCT
of all the capture groups.

Commit:
Merge branch 'master' into regex-struct

Commit:
Issue #6882: REGEXP_EXTRACT Capture Groups

Fix release unused variable.

Tishj commented on src/function/scalar/string/regexp.cpp:
```diff
@@ -218,6 +218,107 @@ bool RegexpExtractBindData::Equals(const FunctionData &other_p) const {
+				ConstantVector::SetNull(*child_entry, false);
+				auto &extracted = ws[col];
+				auto cdata = ConstantVector::GetData<string_t>(*child_entry);
+				cdata[0] = StringVector::AddString(*child_entry, extracted.data(), match ? extracted.size() : 0);
```
You could reuse the string buffer of the input, I've done the same in RegexExtractAll

hawkfish replied:
Ah - interesting!  That was bugging me.

Tishj commented on src/function/scalar/string/regexp.cpp:
```diff
@@ -218,6 +218,107 @@ bool RegexpExtractBindData::Equals(const FunctionData &other_p) const {
+					auto &child_entry = child_entries[col];
+					auto cdata = FlatVector::GetData<string_t>(*child_entry);
+					auto &extracted = ws[col];
+					cdata[i] = StringVector::AddString(*child_entry, extracted.data(), match ? extracted.size() : 0);
```
Same as before, we don't need to do extra allocations because the groups will just be substrings of the input

Tishj reviewed (commented):
Looks good, just had a comment about an optimization we could do
(could be considered not an optimization as well, in the case where you have a really big string and the groups are only a small portion of the string, then keeping this string alive needlessly might be wasteful)

Commit:
Issue #6882: REGEXP_EXTRACT Capture Memory

Reuse the input string memory instead of allocating new buffers for each group column.

hawkfish commented:
> Looks good, just had a comment about an optimization we could do
> (could be considered not an optimization as well, in the case where you have a really big string and the groups are only a small portion of the string, then keeping this string alive needlessly might be wasteful)

A good point, but I have to say that having [worked on a text shaping engine](https://www.youtube.com/watch?v=hmUgKNtC43I), you don't usually end up with a lot of filler data, so I think reuse is the way to go.

Commit:
Merge branch 'master' into regex-struct

Tishj commented on src/function/scalar/string/regexp.cpp:
```diff
@@ -218,6 +218,112 @@ bool RegexpExtractBindData::Equals(const FunctionData &other_p) const {
+		for (idx_t i = 0; i < count; ++i) {
+			const auto idx = iunified.sel->get_index(i);
+			if (ivalidity.RowIsValid(idx)) {
+				str = idata[idx].GetString();
```
I had a look at the "no string inline" test failure, it might be related to this?
Can you try using `str = CreateStringPiece(idata[idx]);` instead?

hawkfish replied:
I looked at that failure and it wasn't in the one test that used my code (although that does seem to call the other regex functions). But I'll give it a go.

hawkfish replied:
Oh I'm an idiot - I was looking at the test _above_ the crash. Passing now.

Commit:
Issue #6882: REGEXP_EXTRACT Capture CreateStringPiece

Switch to CreateStringPiece to make sure strings get created correctly.

Commit:
Merge branch 'master' into regex-struct

Commit:
Issue #6882: REGEXP_EXTRACT Tidy

Fix formatting from merge.

Commit:
Merge branch 'master' into regex-struct

Commit:
Merge branch 'master' into regex-struct

Commit:
Issue #6882: REGEXP_EXTRACT Safe Cast

Use the new safe casting infrastructure.

Commit:
Merge branch 'master' into regex-struct

Mytherin commented on test/sql/function/string/regex_capture.test:
```diff
@@ -0,0 +1,103 @@
+WITH files AS (
+	SELECT f.*, payload FROM filenames f, range(1000) t(payload)
+), extracted AS (
+	SELECT regexp_extract(filename, 'rundate_(\d+-\d+-\d+)_pass_(\d+)', ['rundate', 'pass']) AS groups, payload 
```
Can we add a test with empty/optional groups? (e.g. `rundate_(\d+)?)

Mytherin commented on src/function/scalar/string/regexp.cpp:
```diff
@@ -218,6 +218,111 @@ bool RegexpExtractBindData::Equals(const FunctionData &other_p) const {
+		result.SetVectorType(VectorType::FLAT_VECTOR);
+		auto &res_valid = FlatVector::Validity(result);
+		if (res_valid.GetData()) {
+			res_valid.SetAllValid(count);
```
I don't quite follow why this is required - all result validity values should always be valid initially, no?

hawkfish replied:
Think I just copied it from somewhere else. `DATE_PART` or maybe the other regex code.

hawkfish replied:
I think there used to be problems with structs and their children. Do I not need to reset the children too?

Mytherin replied:
They should be reset correctly as well

Mytherin reviewed (commented):
Thanks for the PR! Looks good. Some comments:

hawkfish commented on src/function/scalar/string/regexp.cpp:
```diff
@@ -227,52 +332,62 @@ static unique_ptr<FunctionData> RegexExtractBind(ClientContext &context, ScalarF
+			if (list_children.empty()) {
+				throw BinderException("%s requires non-empty lists of capture names", bound_function.name);
+			}
+			unordered_set<string> name_collision_set;
```
I was assuming that struct field names are case-sensitive?

Mytherin replied:
They follow the same rules as column names:

```sql
D select {'HELLO': 42, 'hello': 42};
Error: Binder Error: Duplicate struct entry name "hello"
```

Commit:
Merge branch 'master' into regex-struct

Commit:
Issue #6882: REGEXP_EXTRACT Capture Groups

CR feedback: Case-insensitivity, validity masks, more tests.

Commit:
Issue #6882: REGEXP_EXTRACT Capture Groups

More validity masks.

Commit:
Merge branch 'master' into regex-struct

Mytherin commented on test/sql/function/string/regex_capture.test:
```diff
@@ -0,0 +1,143 @@
+2023-02-14	1	3
+
+# Scaled up
+query III
```
There seems to be a missing `rowsort` here for the vsize = 2 tests

Commit:
Merge branch 'master' into regex-struct

Commit:
Issue #6882: REGEXP_EXTRACT Capture Groups

Fix result output for vsize=2

Mytherin commented:
Thanks! LGTM
